File: magellan-sample-advanced/src/main/java/com/wealthfront/magellan/sample/advanced/tide/TideDetailsScreen.java
package com.wealthfront.magellan.sample.advanced.tide;

import android.content.Context;
import android.widget.Toast;
import com.wealthfront.magellan.sample.advanced.NoaaApi;
import com.wealthfront.magellan.sample.advanced.R;
import com.wealthfront.magellan.sample.advanced.base.RefWatcherScreen;
import com.wealthfront.magellan.sample.advanced.model.Observation;
import com.wealthfront.magellan.sample.advanced.model.TideInfo;
import java.math.BigDecimal;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.List;
import javax.inject.Inject;
import rx.functions.Action1;

import static com.wealthfront.magellan.sample.advanced.SampleApplication.app;
import static rx.android.schedulers.AndroidSchedulers.mainThread;

public class TideDetailsScreen extends RefWatcherScreen<TideDetailsView> {

  public static final Comparator<Observation> OBSERVATION_COMPARATOR = new Comparator<Observation>() {
    @Override
    public int compare(Observation o1, Observation o2) {
      return o1.getVerifiedWaterLevel().compareTo(o2.getVerifiedWaterLevel());
    }
  };

  @Inject NoaaApi noaaApi;
  private final String tideLocationName;
  int noaaApiId;

  TideDetailsScreen(int noaaApiId, String tideLocationName) {
    this.noaaApiId = noaaApiId;
    this.tideLocationName = tideLocationName;
  }

  @Override
  protected TideDetailsView createView(Context context) {
    app(context).injector().inject(this);
    return new TideDetailsView(context);
  }

  @Override
  public String getTitle(Context context) {
    return tideLocationName;
  }

  @Override
  protected void onSubscribe(final Context context) {
    autoUnsubscribe(noaaApi.getTideInfo(noaaApiId)
        .observeOn(mainThread())
        .subscribe(new Action1<TideInfo>() {
          @Override
          public void call(TideInfo tideInfo) {
            if (tideInfo.getData() != null && !tideInfo.getData().isEmpty()) {
              List<Observation> observations = tideInfo.getData();
              BigDecimal highestMeasuredTideHeight =
                  Collections.max(filterOutNullMeasurements(observations), OBSERVATION_COMPARATOR)
                      .getVerifiedWaterLevel();
              BigDecimal lowestMeasuredTideHeight =
                  Collections.min(filterOutNullMeasurements(observations), OBSERVATION_COMPARATOR)
                      .getVerifiedWaterLevel();
              BigDecimal latestMeasuredTideHeight =
                  observations.get(observations.size() - 1).getVerifiedWaterLevel();
              getView().setTideHeights(latestMeasuredTideHeight, lowestMeasuredTideHeight,
                  highestMeasuredTideHeight);
            }
          }
        }, new Action1<Throwable>() {
          @Override
          public void call(Throwable throwable) {
            Toast.makeText(context, R.string.cannot_retrieve_tide_info, Toast.LENGTH_SHORT).show();
          }
        }));
  }

  private static List<Observation> filterOutNullMeasurements(List<Observation> listWithNulls) {
    List<Observation> result = new ArrayList<>();
    for (Observation item : listWithNulls) {
      if (item.getVerifiedWaterLevel() != null) {
        result.add(item);
      }
    }
    return result;
  }
}

File: magellan-sample-advanced/src/main/java/com/wealthfront/magellan/sample/advanced/tide/TideDetailsScreen.kt
package com.wealthfront.magellan.sample.advanced.tide

import android.animation.Animator
import android.animation.AnimatorListenerAdapter
import android.content.Context
import android.view.View
import android.view.ViewGroup
import android.widget.FrameLayout
import android.widget.LinearLayout
import android.widget.ProgressBar
import android.widget.TextView
import android.widget.Toast
import androidx.annotation.VisibleForTesting
import com.wealthfront.magellan.compose.core.Screen
import com.wealthfront.magellan.sample.advanced.NoaaApi
import com.wealthfront.magellan.sample.advanced.R
import com.wealthfront.magellan.sample.advanced.SampleApplication.Companion.app
import com.wealthfront.magellan.sample.advanced.databinding.TideDetailBinding
import com.wealthfront.magellan.sample.advanced.model.Observation
import rx.android.schedulers.AndroidSchedulers
import java.math.BigDecimal
import java.math.RoundingMode
import java.util.ArrayList
import java.util.Collections
import java.util.Comparator
import java.util.Locale
import javax.inject.Inject

class TideDetailsScreen(var noaaApiId: Int) : Screen<TideDetailBinding>(TideDetailBinding::inflate) {

  @Inject lateinit var noaaApi: NoaaApi

  override fun onCreate(context: Context) {
    app(context).injector().inject(this)
  }

  override fun onShow(context: Context, binding: TideDetailBinding) {
    noaaApi.getTideInfo(noaaApiId)
      .observeOn(AndroidSchedulers.mainThread())
      .subscribe({ tideInfo ->
        if (tideInfo.data != null && tideInfo.data.isNotEmpty()) {
          val observations = tideInfo.data
          val highestMeasuredTideHeight = Collections.max(
            filterOutNullMeasurements(observations),
            OBSERVATION_COMPARATOR)
            .verifiedWaterLevel
          val lowestMeasuredTideHeight = Collections.min(
            filterOutNullMeasurements(observations),
            OBSERVATION_COMPARATOR)
            .verifiedWaterLevel
          val latestMeasuredTideHeight =
            observations[observations.size - 1].verifiedWaterLevel
          setTideHeights(
            binding,
            latestMeasuredTideHeight, lowestMeasuredTideHeight,
            highestMeasuredTideHeight)
        }
      }, {
        Toast.makeText(context, R.string.cannot_retrieve_tide_info, Toast.LENGTH_SHORT).show()
      })
  }

  @VisibleForTesting
  fun setTideHeights(
    binding: TideDetailBinding,
    latestMeasuredTideHeight: BigDecimal, lowestMeasuredTideHeight: BigDecimal,
    highestMeasuredTideHeight: BigDecimal
  ) {
    showContent(binding)
    displayWaterLevelText(
      binding,
      latestMeasuredTideHeight, lowestMeasuredTideHeight,
      highestMeasuredTideHeight)
    setVisibleWaterLevel(
      binding,
      latestMeasuredTideHeight, lowestMeasuredTideHeight,
      highestMeasuredTideHeight)
  }

  private fun showContent(binding: TideDetailBinding) {
    binding.tideDetailsContent.alpha = 0f
    binding.tideDetailsContent.visibility = View.VISIBLE
    binding.loading.animate().alpha(0f).setListener(object : AnimatorListenerAdapter() {
      override fun onAnimationEnd(animation: Animator) {
        binding.tideDetailsContent.animate().alpha(1f).start()
        binding.loading.visibility = View.GONE
      }
    }).start()
  }

  private fun displayWaterLevelText(
    binding: TideDetailBinding,
    latestMeasuredTideHeight: BigDecimal, lowestMeasuredTideHeight: BigDecimal,
    highestMeasuredTideHeight: BigDecimal
  ) {
    binding.currentWaterLevel.text = String.format(
      Locale.US, "Current Water Level: %.2f ft",
      latestMeasuredTideHeight.setScale(2, RoundingMode.HALF_UP))
    binding.highestWaterLevel.text = String.format(
      Locale.US, "Today's Highest Water Level: %.2f ft",
      highestMeasuredTideHeight.setScale(2, RoundingMode.HALF_UP))
    binding.lowestWaterLevel.text = String.format(
      Locale.US, "Today's Lowest Water Level: %.2f ft",
      lowestMeasuredTideHeight.setScale(2, RoundingMode.HALF_UP))
  }

  private fun setVisibleWaterLevel(
    binding: TideDetailBinding,
    latestMeasuredTideHeight: BigDecimal, lowestMeasuredTideHeight: BigDecimal,
    highestMeasuredTideHeight: BigDecimal
  ) {
    val percentOfMax: Float = latestMeasuredTideHeight.subtract(lowestMeasuredTideHeight)
      .divide(
        highestMeasuredTideHeight.subtract(lowestMeasuredTideHeight),
        RoundingMode.HALF_UP)
      .multiply(BigDecimal.valueOf(100))
      .setScale(0, BigDecimal.ROUND_HALF_UP)
      .toFloat()
    binding.currentWaterLevelBottomSpacing.layoutParams = LinearLayout.LayoutParams(
      ViewGroup.LayoutParams.MATCH_PARENT, 2, percentOfMax)
    binding.currentWaterLevelTopSpacing.layoutParams = LinearLayout.LayoutParams(
      ViewGroup.LayoutParams.MATCH_PARENT, 2, 100 - percentOfMax)
  }

  companion object {
    val OBSERVATION_COMPARATOR =
      Comparator<Observation> { o1, o2 -> o1.verifiedWaterLevel.compareTo(o2.verifiedWaterLevel) }

    private fun filterOutNullMeasurements(listWithNulls: List<Observation>): List<Observation> {
      val result: MutableList<Observation> = ArrayList()
      for (item in listWithNulls) {
        if (item.verifiedWaterLevel != null) {
          result.add(item)
        }
      }
      return result
    }
  }
}